lib/wikitrivia/game.ex

```diff
@@ -1,100 +1,229 @@
+  end
+
+  # TODO: We'll probably want to alter the behavior here s.t. the final player answering the
+  # question immediately triggers the next phase (the question_results) phase. (there's no reason
+  # to wait those extra seconds, since you can't change your answer - OR maybe we let people change
+  # their answers after everyone has answered).
+
+  # Starts the pose question -> show results loop. Ends after current_question reaches
+  # num_questions - 1. The provided game_timer_callback function will be called after every timed
+  # state change.
+
+  # THIS EVENT CALL DOES NOT RETURN ANY STATE (the provided callback function should be relied upon for that)
   def start(game_id, game_timer_callback) do
-    start_question(game_id, game_timer_callback)
+    GenServer.cast(name_by_game_id(game_id), {:start_game, game_timer_callback})
   end
 
-  # State structure of a Game. Details:
-  #   * players - Set of all players in this game.
-  #   * scores - Map of all player scores in this game, keyed by player name
-  #   * num_questions_left - Number of questions remaining to answer in this game
-  #   * timer_state - Current state of timed component of this game. :off before entering timed
-  #       periods, otherwise the name of the period (:question, :question_results, :done)
-  #   * timer_data - Data associated with the timer_state (e.g. the question)
-  defp default_state do
-    %{
-      players: MapSet.new(),
-      scores: %{},
-      num_questions_left: 5,
-      timer_state: :off,
-      timer_data: %{}
-    }
+
+  ### Server (callbacks) ###
+
+  def init(state) do
+    {:ok, state}
   end
 
-  defp add_player(player_name) do
-    fn (state = %{players: players, scores: scores}) ->
-      players = players |> MapSet.put(player_name)
-      scores = scores |> Map.put(player_name, 0)
-      %{state | players: players, scores: scores}
+  def handle_call(:get_state, _from, state), do: {:reply, state, state}
+
+  def handle_call({:add_player, player_id}, _from, state = %{players: players, scores: scores}) do
+    if Map.has_key?(players, player_id) do
+      {:reply, {:no_change, state}, state}
+    else
+      new_state = %{state |
+        players: Map.put(players, player_id, @default_player_name),
+        scores: Map.put(scores, player_id, 0)
+      }
+      {:reply, {:ok, new_state}, new_state}
     end
   end
 
-  defp award_points(player_name, points) do
-    fn (state = %{scores: scores}) ->
-      %{^player_name => player_score} = scores
-      new_score = player_score + points
-      %{state | scores: %{scores | player_name => new_score}}
+  def handle_call({:update_player_name, player_id, player_name}, _from, state = %{players: players}) do
+    cond do
+      !Map.has_key?(players, player_id) ->
+        {:reply, {:no_change, state}, state}
+      Map.fetch(players, player_id) == {:ok, player_name} ->
+        {:reply, {:no_change, state}, state}
+      true ->
+        new_state = %{state |
```

Comment:
@denheck I don't think so - I don't think that would handle the situation where player_id is not present correctly. This code will bork if player_id isn't already in the `players` map:
```elixir
%{players | player_id => player_name} 
```
And this will add a new player if they aren't in the map, which I don't think is the desired behavior (we wanna ignore player_ids that aren't in the players list here - i.e. no updating names of players that weren't part of a previous add_player call)
```elixir
Map.put(players, player_id, player_name)
```

That said, more generally, I bet we could reduce some of the no_changes to that sort of thing. I'm not familiar enough with elixir to know if there's a way to only set a value on a map if the key already exists, but otherwise to do nothing - which is what we'd need in this situation to handle the first `cond` clause.